openmls/src/framing/plaintext.rs

```diff
@@ -49,20 +35,82 @@ use tls_codec::{Serialize, TlsByteVecU32, TlsDeserialize, TlsSerialize, TlsSize}
 #[derive(Debug, PartialEq, Clone, Serialize, Deserialize, TlsSerialize, TlsSize)]
 pub(crate) struct MlsPlaintext {
     wire_format: WireFormat,
-    group_id: GroupId,
-    epoch: GroupEpoch,
-    sender: Sender,
-    authenticated_data: TlsByteVecU32,
-    content_type: ContentType,
-    content: MlsPlaintextContentType,
+    content: MlsContent,
     signature: Signature,
     confirmation_tag: Option<ConfirmationTag>,
     membership_tag: Option<MembershipTag>,
 }
 
+/// ```c
+/// // draft-ietf-mls-protocol-16
+/// struct {
+///     opaque group_id<V>;
+///     uint64 epoch;
+///     Sender sender;
+///     opaque authenticated_data<V>;
+///
+///     // ... continued in [MlsContentBody] ...
+/// } MLSContent;
+/// ```
+#[derive(
+    Debug, PartialEq, Clone, Serialize, Deserialize, TlsSerialize, TlsDeserialize, TlsSize,
+)]
+pub(crate) struct MlsContent {
+    pub(crate) group_id: GroupId,
+    pub(crate) epoch: GroupEpoch,
+    pub(crate) sender: Sender,
+    pub(crate) authenticated_data: TlsByteVecU32,
+
+    pub(crate) body: MlsContentBody,
+}
+
+/// ```c
+/// // draft-ietf-mls-protocol-16
+/// struct {
+///     // ... continued from [MlsContent] ...
+///
+///     ContentType content_type;
+///     select (MLSContent.content_type) {
+///         case application:
+///           opaque application_data<V>;
+///         case proposal:
+///           Proposal proposal;
+///         case commit:
+///           Commit commit;
+///     }
+/// } MLSContent;
+/// ```
+#[derive(
+    Debug, PartialEq, Clone, Serialize, Deserialize, TlsSerialize, TlsDeserialize, TlsSize,
+)]
+#[repr(u8)]
+pub(crate) enum MlsContentBody {
+    #[tls_codec(discriminant = 1)]
+    Application(TlsByteVecU32),
+    #[tls_codec(discriminant = 2)]
+    Proposal(Proposal),
+    #[tls_codec(discriminant = 3)]
+    Commit(Commit),
+}
+
+impl MlsContentBody {
+    pub(crate) fn content_type(&self) -> ContentType {
+        match self {
+            Self::Application(_) => ContentType::Application,
+            Self::Proposal(_) => ContentType::Proposal,
+            Self::Commit(_) => ContentType::Commit,
+        }
+    }
+}
+
+impl From<MlsPlaintext> for MlsContentBody {
+    fn from(plaintext: MlsPlaintext) -> Self {
+        plaintext.content.body
+    }
+}
+
 pub(crate) struct Payload {
-    pub(crate) payload: MlsPlaintextContentType,
-    pub(crate) content_type: ContentType,
+    pub(crate) payload: MlsContentBody,
```

Comment:
With the payload being only the `MlsContentBody` now we don't need it anymore.

Reply:
Will do. I wanted to keep the changes small. But I guess this still fits in.